abuse/signup.md

```diff
@@ -46,8 +46,81 @@ channels are in English.
+The design criteria for it are as follows:
 
-XXXTODO: Describe the required API changes.
+1. change as little as possible the current signup API;
+2. be stateless, in conformance with requirement 4;
+3. be secure, in the following ways:
+   - allow a limited (configurable) time to solve the CAPTCHA;
+   - prevent users from reusing CAPTCHA answers;
+   - do not expose any information to the user that may facilitate
+	 automated CAPTCHA solving;
+4. be independent from [TextCaptcha]: the CAPTCHA-generating system
+   must be replaceable without any change to the API.
+
+
+#### Public API
+
+A `/captcha` endpoint is added, expecting a JSON object with a single
+`username` attribute.  The reply contains:
+
+- a `challenge` string, the human-readable question;
+- an opaque `token`, serialized as a string;
+- an `expiration` time, serialized as an integer timestamp.
+
+When using the `/user/create` endpoint for user creation, the client
+must provide (in addition to the current requirements):
+
+- the opaque `token` it received from a previous call to `answer`,
+  for the `username` it is requesting;
+- a matching `answer` string.
+
+The `/user/create` implementation must perform all other validation
+checks (and error-out accordingly) before validating the CAPTCHA.
+Doing otherwise would expose an interactive verifier for the CAPTCHA
+solution (which might or might not be an exploitable flaw).
+
+
+#### Opaque token
+
+The CAPTCHA validation requires three pieces of information:
+
+- the `a` value returned by [TextCaptcha];
+- the expiration `timestamp`;
+- the `username` requested during CAPTCHA generation.
+
+This data needs to be integrity-protected, since an attacker able to modify
+any of its parts would be able to violate the security requirements (3) or
+bypass the CAPTCHA entirely.
+
+Moreover, it needs to be kept confidential: the `a` attribute is a hash of
+the valid, lowercased answers: exposing it to the user reveals a *verifier*
+for the valid answers, enabling a malicious user to bruteforce them offline.
+
+As such, the `token` opaque value is generated as follows:
+
+- the required data (`a`, `timestamp` and `username`) is serialized
+  using an implementation-defined mechanism.
+  [Snappy](https://github.com/golang/snappy)-compressed JSON is suitable.
+- the serialized data is encrypted, using an authenticated encryption
+  primitive such as AES-GCM, against a constant, symmetric key that is
+  randomly generated when the API server starts (and never persisted to
+  disk);
+- the encrypted data is Base64-encoded, using the
+  [RFC 4648 URL-safe alphabet](https://tools.ietf.org/html/rfc4648#section-5).
+
+The use of a random, volatile key for token encryption implies two trade-offs:
+
+- CAPTCHA tokens are implicitely expired when the application is restarted;
+  they do expire within a short timeframe anyway, mitigating the issue;
+- `api.hashbang.sh` can only be served by a single instance.
```

Comment:
There is nothing with this design that prevents > 1 instances. We would simply employ sticky-sessions on the load balance.